src/hera/workflows/_mixins.py

```diff
@@ -480,42 +480,16 @@ def __call__(self, *args, **kwargs) -> Optional[SubNodeMixin]:
-            kwargs_arguments = (
-                kwargs_arguments if isinstance(kwargs_arguments, list) else [kwargs_arguments]
-            )  # type: ignore
-            arguments = (
-                self.arguments if isinstance(self.arguments, list) else [self.arguments] + kwargs_arguments
-            )  # type: ignore
-            arguments = list(filter(lambda x: x is not None, arguments))
-            parameters = [arg for arg in arguments if isinstance(arg, ModelParameter) or isinstance(arg, Parameter)]
-            parameter_names = {p.name for p in parameters}
-            artifacts = [arg for arg in arguments if isinstance(arg, ModelArtifact) or isinstance(arg, Artifact)]
-            artifact_names = {a.name for a in artifacts}
+            arguments = self._get_arguments(**kwargs)
+            parameter_names = self._get_parameter_names(arguments)
+            artifact_names = self._get_artifact_names(arguments)
             if "source" in kwargs and "with_param" in kwargs:
-                # Argo uses the `inputs` field to indicate the expected parameters of a specific template whereas the
-                # `arguments` are used to indicate exactly what _values_ are assigned to the set inputs. Here,
-                # we infer the arguments when `with_param` is used. If a source is passed along with `with_param`, we
-                # infer the arguments to set from the given source. It is assumed that `with_param` will return the
-                # expected result for Argo to fan out the task on
-                new_parameters = _get_params_from_source(kwargs["source"])
-                if new_parameters is not None:
-                    for p in new_parameters:
-                        if p.name not in parameter_names and p.name not in artifact_names:
-                            arguments.append(p)
+                arguments += self._get_deduped_params_from_source(
+                    parameter_names, artifact_names, kwargs["source"]
+                )
             elif "source" in kwargs and "with_items" in kwargs:
-                # similarly to the above, we can infer the arguments to create based on the content of `with_items`.
-                # The main difference between `with_items` and `with_param` is that param is a serialized version of
-                # `with_items`. Hence, `with_items` comes in the form of a list of objects, whereas `with_param` comes
-                # in as a single serialized object. Here, we can infer the parameters to create based on the content
-                # of `with_items`
-                new_parameters = _get_params_from_items(kwargs["with_items"])
-                if new_parameters is not None:
-                    for p in new_parameters:
-                        if p.name not in parameter_names:
-                            arguments.append(p)
+                arguments = arguments + self._get_deduped_params_from_items(parameter_names, kwargs["with_items"])
```

Comment:
To match `if` branch above
```suggestion
                arguments += self._get_deduped_params_from_items(parameter_names, kwargs["with_items"])
```